File: 649-dota2-senate/dota2-senate.cpp
#include <iostream>
#include <queue>
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> r_queue;
        queue<int> d_queue;
        int n = senate.size();
        for(int i = 0;i<n;i++){
            if (senate[i] == 'R'){
                r_queue.push(i);
            }
            else{
                d_queue.push(i);
            }
        }
        while(!r_queue.empty() && !d_queue.empty()){
            int r = r_queue.front();
            int d = d_queue.front();
            d_queue.pop();
            r_queue.pop();
            ///////////////////
            if (r<d){
                r_queue.push(r+n);
            }
            else{
                d_queue.push(d+n);
            }
        }
        return d_queue.empty() ? "Radiant" :"Dire";
    }
};

File: 649-dota2-senate/dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        queue = list(senate)
        #ret = ''
        r_count=d_count=r_ban=d_ban = 0
        ##################
        for s in senate:
            if s =="R":
                r_count+=1
            else:
                d_count+=1
        while(queue):
            curr = queue.pop(0)
            if curr == "R":
                if r_ban>0:
                    r_ban-=1
                else:
                    d_ban+=1
                    queue.append(curr)
            else:
                if d_ban>0:
                    d_ban-=1
                else:
                    r_ban+=1
                    queue.append(curr)
            if r_count<=r_ban:
                return "Dire"
            elif d_count<=d_ban:
                return "Radiant"
        #########################
